### PR TITLE
[megatron] feat: Support Megatron chunk entropy

### PR DESCRIPTION
**Pull request by ZLiao097**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

Currently, chunk entropy is only available for the FSDP training backend in VERL and not yet implemented for Megatron. This feature significantly cuts peak GPU memory usage in long-sequence scenarios. In the Qwen3.5-35B Megatron test with sequence length expanding from 2k to 20k, peak memory consumption drops by 8.6GB.
Before chunk entropy:
image

After chunk entropy:
image

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/utils/megatron/tensor_parallel.py [hunk 1]
```diff
@@ -151,6 +151,23 @@ def vocab_parallel_entropy(vocab_parallel_logits: torch.Tensor) -> torch.Tensor:
...
+    Returns: (batch_size, seq_len)
+    """
+    
+    output_shape = list(vocab_parallel_logits.shape[:-1])
+    entropy = torch.zeros(output_shape, device=vocab_parallel_logits.device)
+    for i in range(0, vocab_parallel_logits.shape[1], chunk_size):
+        logits_chunk = vocab_parallel_logits[:, i : i + chunk_size, :]
+        entropy_chunk = _VocabParallelEntropy.apply(logits_chunk)
+        entropy[:, i : i + chunk_size] = entropy_chunk
+    return entropy
```
[the remark this answers is not in the document]
Comment (ZLiao097) [reply verl/utils/megatron/tensor_parallel.py:170]: 1. The logits passed to `vocab_parallel_entropy` are always 3D because they come directly from Megatron's `ColumnParallelLinear` (LM Head), which outputs 3D tensors:

- **THD**: `(1, total_nnz, V//tp)` — all sequences packed into B=1
- **BSHD**: `(B, max_seqlen, V//tp)` — padded to max sequence length
So it is unnecessary for current usage where inputs are always 3D.
 
2.  Follow the dtype convention used in FSDP's chunk entropy implementation `entropy_from_logits_with_chunking` : keep it consistent with the dtype of the input logits.

File: verl/utils/megatron/tensor_parallel.py [hunk 2]
```diff
@@ -151,6 +151,23 @@ def vocab_parallel_entropy(vocab_parallel_logits: torch.Tensor) -> torch.Tensor:
     return _VocabParallelEntropy.apply(vocab_parallel_logits)
 
 
+def vocab_parallel_entropy_with_chunking(vocab_parallel_logits: torch.Tensor, chunk_size: int = 2048) -> torch.Tensor:
```
Comment (ETOgaosion) [inline verl/utils/megatron/tensor_parallel.py:154]: Thanks for contribution! This is another moderate way to handle memory peak. Shall we let chunk_size editable by config?

Comment (ZLiao097) [reply verl/utils/megatron/tensor_parallel.py:154]: Got it. I will supplement the commit shortly.

**Comment by ZLiao097** [comment]
@ETOgaosion plz review again

File: verl/workers/engine/fsdp/transformer_impl.py [hunk 3]
```diff
@@ -1124,7 +1096,10 @@ def prepare_model_outputs(self, output, output_args, micro_batch: TensorDict, lo
                 # compute entropy
                 if calculate_entropy:
                     if not self.engine_config.entropy_checkpointing:
-                        entropy_rmpad = self.compute_entropy_from_logits(logits_rmpad)  # ((total_nnz / sp) + pad)
+                        entropy_rmpad = self.compute_entropy_from_logits(
+                            logits_rmpad,
+                            chunk_size=self.engine_config.entropy_from_logits_chunk_size,
```
Comment (zyang6) [inline verl/workers/engine/fsdp/transformer_impl.py]: Init already selects entropy_from_logits_with_chunking or entropy_from_logits (L149–158). At the call site, chunk_size is always passed. When chunking is disabled, this may raise TypeError since entropy_from_logits only accepts logits.

Suggestion: either keep the call as compute_entropy_from_logits(logits_rmpad) and bind chunk_size via functools.partial at init, or add an if at the call site to only pass chunk_size when chunking is enabled.

File: verl/utils/megatron/tensor_parallel.py [hunk 4]
```diff
@@ -151,6 +151,25 @@ def vocab_parallel_entropy(vocab_parallel_logits: torch.Tensor) -> torch.Tensor:
...
+    """Memory-efficient entropy calculation using chunked processing when logits are sharded in tp ranks.
+    Args:
+        vocab_parallel_logits: (batch_size, seq_len, vocab_size // tp_size) or (total_nnz, vocab_size // tp_size)
+        chunk_size: Number of sequence tokens to process at once. Defaults to 2048.
+    Returns: (batch_size, seq_len)
+    """
+
+    output_shape = list(vocab_parallel_logits.shape[:-1])
+    entropy = torch.zeros(output_shape, device=vocab_parallel_logits.device)
```
Comment (duesdues) [inline verl/utils/megatron/tensor_parallel.py:163]: dtype mismatch: torch.zeros(output_shape, device=...) defaults to float32, while logits are typically bf16/fp16. Use dtype=vocab_parallel_logits.dtype (or torch.empty + slice assignment).

File: verl/workers/engine/fsdp/transformer_impl.py (hunk 3 above)
Comment (ZLiao097) [reply verl/workers/engine/fsdp/transformer_impl.py]: thx, resolve~

**Review by ETOgaosion — APPROVED** [review]
